src/transformers/clip.rs

```diff
@@ -313,29 +353,29 @@ impl<B: Backend> ClipTextTransformer<B> {
     }
 
     // https://github.com/huggingface/transformers/blob/674f750a57431222fa2832503a108df3badf1564/src/transformers/models/clip/modeling_clip.py#L678
-    fn build_causal_attention_mask(bsz: usize, seq_len: usize, device: &Backend::Device) -> Tensor<B, 3> {
+    fn build_causal_attention_mask(bsz: usize, seq_len: usize, device: &B::Device) -> Tensor<B, 2> {
         let mask = Tensor::full_device([bsz, seq_len, seq_len], f32::MIN, device);
         let mask = zero_lower_diagonal(mask); // zero out the lower diagonal
         let mask = mask.unsqueeze(); // expand mask
         mask
     }
 
     fn forward(&self, xs: Tensor<B, 2, Int>) -> Tensor<B, 3> {
-        let [bsz, seq_len, _] = xs.dims();
-        let xs = self.embeddings.forward(xs)?;
-        let causal_attention_mask = Self::build_causal_attention_mask(bsz, seq_len, xs.device());
-        let xs = self.encoder.forward(xs, &causal_attention_mask)?;
+        let [bsz, seq_len] = xs.dims();
+        let xs = self.embeddings.forward(xs);
+        let causal_attention_mask = Self::build_causal_attention_mask(bsz, seq_len, &xs.device());
+        let xs = self.encoder.forward(xs, &causal_attention_mask);
         self.final_layer_norm.forward(xs)
     }
 }
 
-fn zero_lower_diagonal<B: Backend>(mut xs: Tensor<B, 3>) -> Tensor<B, 3> {
-    let [bsz, seq_len, _] = xs.dims();
-    for i in 0..seq_len {
-        for j in 0..i {
-            xs[[i, j, 0]] = 0.0; // Assuming the third dimension is the channel/diagonal dimension
-        }
-    }
+fn zero_lower_diagonal<B: Backend>(xs: Tensor<B, 3>) -> Tensor<B, 3> {
+    let [m, n, _] = xs.dims();
+
+    // build an upper-triangle matrix
+    let upper_diag = (0..max(m, n))
+        .map(Tensor::<B, 2, Int>::diagonal)
+        .fold(Tensor::zeros([max(m, n); 2]), Tensor::add);
 
-    xs
+    upper_diag.reshape([m, n]).unsqueeze().float().mul(xs)
```

Comment:
The reshape might mess up the triangle. I don't know if we can assume that m == n and the way reshape works could shift some values the wrong way